File: src-tauri/src/commands/transactions.rs
use crate::db::queries;
use crate::models::{AssetHoldingSummary, Transaction, TxType};
use crate::state::AppState;
use tauri::State;

#[tauri::command]
pub fn add_transaction(
    asset_id: String,
    tx_type: String,
    quantity: f64,
    price_usd: f64,
    ts: i64,
    notes: Option<String>,
    state: State<'_, AppState>,
) -> Result<Transaction, String> {
    let tx_type = TxType::from_str(&tx_type).map_err(|e| e.to_string())?;
    state
        .with_db(|conn| {
            queries::transactions::insert_transaction(
                conn,
                &asset_id,
                &tx_type,
                quantity,
                price_usd,
                ts,
                notes.as_deref(),
            )
        })
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn list_transactions(
    asset_id: String,
    state: State<'_, AppState>,
) -> Result<Vec<Transaction>, String> {
    state
        .with_db(|conn| queries::transactions::list_transactions_by_asset(conn, &asset_id))
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn delete_transaction(
    id: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    state
        .with_db(|conn| queries::transactions::soft_delete_transaction(conn, &id))
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_holding_summary(
    asset_id: String,
    state: State<'_, AppState>,
) -> Result<AssetHoldingSummary, String> {
    state
        .with_db(|conn| queries::transactions::get_holding_summary(conn, &asset_id))
        .map_err(|e| e.to_string())
}

File: src/lib/tauri/transactions.ts
import { invoke } from "@tauri-apps/api/core";
import type { Transaction, TxType, AssetHoldingSummary } from "@/types";

export async function addTransaction(
  assetId: string,
  txType: TxType,
  quantity: number,
  priceUsd: number,
  ts: number,
  notes?: string,
): Promise<Transaction> {
  return invoke<Transaction>("add_transaction", {
    assetId,
    txType,
    quantity,
    priceUsd,
    ts,
    notes: notes ?? null,
  });
}

export async function listTransactions(
  assetId: string,
): Promise<Transaction[]> {
  return invoke<Transaction[]>("list_transactions", { assetId });
}

export async function updateTransaction(
  id: string,
  txType: TxType,
  quantity: number,
  priceUsd: number,
  ts: number,
  notes?: string,
): Promise<void> {
  return invoke<void>("update_transaction", {
    id,
    txType,
    quantity,
    priceUsd,
    ts,
    notes: notes ?? null,
  });
}

export async function deleteTransaction(id: string): Promise<void> {
  return invoke<void>("delete_transaction", { id });
}

export async function getHoldingSummary(
  assetId: string,
): Promise<AssetHoldingSummary> {
  return invoke<AssetHoldingSummary>("get_holding_summary", { assetId });
}
